// Raft Consensus Algorithm gRPC Service Definition
// Based on the Raft paper: https://raft.github.io/raft.pdf
//
// This proto defines the RPC interface that candidates must implement
// to build a working Raft consensus cluster.

syntax = "proto3";

package raft;

option go_package = "github.com/sdp/raft";
option java_package = "com.sdp.raft";
option java_multiple_files = true;

// ============================================================================
// Core Raft RPCs (Required for basic consensus)
// ============================================================================

// RaftService defines the core Raft consensus RPCs
service RaftService {
    // RequestVote is invoked by candidates to gather votes during elections.
    // Candidates use this RPC to request votes from other servers when
    // they start an election.
    rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);

    // AppendEntries is invoked by leader to replicate log entries and
    // also used as heartbeat. Leaders send this RPC to all followers
    // to replicate log entries and maintain authority.
    rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);

    // InstallSnapshot is invoked by leader to send chunks of a snapshot
    // to a follower that is too far behind.
    rpc InstallSnapshot(InstallSnapshotRequest) returns (InstallSnapshotResponse);
}

// ============================================================================
// Client-facing Key-Value Store Service
// ============================================================================

// KeyValueService provides a simple key-value store interface for clients.
// The Raft cluster should reach consensus on these operations.
service KeyValueService {
    // Get retrieves a value for the given key
    rpc Get(GetRequest) returns (GetResponse);

    // Put stores a key-value pair (requires consensus)
    rpc Put(PutRequest) returns (PutResponse);

    // Delete removes a key (requires consensus)
    rpc Delete(DeleteRequest) returns (DeleteResponse);

    // GetLeader returns the current leader information
    rpc GetLeader(GetLeaderRequest) returns (GetLeaderResponse);

    // GetClusterStatus returns the cluster health status
    rpc GetClusterStatus(GetClusterStatusRequest) returns (GetClusterStatusResponse);
}

// ============================================================================
// Raft Core Message Types
// ============================================================================

// LogEntry represents a single entry in the Raft log
message LogEntry {
    // Index of this entry in the log (1-indexed)
    uint64 index = 1;

    // Term when entry was received by leader
    uint64 term = 2;

    // Command to be applied to the state machine
    bytes command = 3;

    // Type of command (e.g., "put", "delete")
    string command_type = 4;
}

// RequestVoteRequest is sent by candidates to request votes
message RequestVoteRequest {
    // Candidate's term
    uint64 term = 1;

    // Candidate requesting vote
    string candidate_id = 2;

    // Index of candidate's last log entry
    uint64 last_log_index = 3;

    // Term of candidate's last log entry
    uint64 last_log_term = 4;
}

// RequestVoteResponse is the response to RequestVote
message RequestVoteResponse {
    // Current term, for candidate to update itself
    uint64 term = 1;

    // True means candidate received vote
    bool vote_granted = 2;
}

// AppendEntriesRequest is sent by leader to replicate log entries
message AppendEntriesRequest {
    // Leader's term
    uint64 term = 1;

    // Leader's ID so follower can redirect clients
    string leader_id = 2;

    // Index of log entry immediately preceding new ones
    uint64 prev_log_index = 3;

    // Term of prev_log_index entry
    uint64 prev_log_term = 4;

    // Log entries to store (empty for heartbeat)
    repeated LogEntry entries = 5;

    // Leader's commit index
    uint64 leader_commit = 6;
}

// AppendEntriesResponse is the response to AppendEntries
message AppendEntriesResponse {
    // Current term, for leader to update itself
    uint64 term = 1;

    // True if follower contained entry matching prev_log_index and prev_log_term
    bool success = 2;

    // For faster log replication: the follower's last log index
    // (helps leader find the right match point faster)
    uint64 match_index = 3;
}

// InstallSnapshotRequest is sent by leader to transfer a snapshot
message InstallSnapshotRequest {
    // Leader's term
    uint64 term = 1;

    // Leader's ID
    string leader_id = 2;

    // The snapshot replaces all entries up through and including this index
    uint64 last_included_index = 3;

    // Term of last_included_index
    uint64 last_included_term = 4;

    // Byte offset where chunk is positioned in the snapshot file
    uint64 offset = 5;

    // Raw bytes of the snapshot chunk
    bytes data = 6;

    // True if this is the last chunk
    bool done = 7;
}

// InstallSnapshotResponse is the response to InstallSnapshot
message InstallSnapshotResponse {
    // Current term, for leader to update itself
    uint64 term = 1;
}

// ============================================================================
// Key-Value Store Message Types
// ============================================================================

message GetRequest {
    string key = 1;
}

message GetResponse {
    string value = 1;
    bool found = 2;
    string error = 3;
}

message PutRequest {
    string key = 1;
    string value = 2;
}

message PutResponse {
    bool success = 1;
    string error = 2;
    // Leader hint if this node is not the leader
    string leader_hint = 3;
}

message DeleteRequest {
    string key = 1;
}

message DeleteResponse {
    bool success = 1;
    string error = 2;
    string leader_hint = 3;
}

message GetLeaderRequest {}

message GetLeaderResponse {
    string leader_id = 1;
    string leader_address = 2;
    bool is_leader = 3;
}

message GetClusterStatusRequest {}

message GetClusterStatusResponse {
    // Current node's state
    string node_id = 1;
    string state = 2;  // "follower", "candidate", "leader"
    uint64 current_term = 3;
    string voted_for = 4;
    uint64 commit_index = 5;
    uint64 last_applied = 6;

    // Log info
    uint64 log_length = 7;
    uint64 last_log_term = 8;

    // Cluster members
    repeated ClusterMember members = 9;
}

message ClusterMember {
    string node_id = 1;
    string address = 2;
    string state = 3;  // Known state of this member
    uint64 match_index = 4;  // Only meaningful if this node is leader
    uint64 next_index = 5;   // Only meaningful if this node is leader
}
